FirebaseStorageSwift/Tests/Integration/StorageAsyncAwait.swift

```diff
@@ -0,0 +1,391 @@
+    }
+
+    func testSimplePutDataNoMetadata() async throws {
+      let ref = storage.reference(withPath: "ios/public/testSimplePutDataNoMetadata")
+      let data = try XCTUnwrap("Hello Swift World".data(using: .utf8), "Data construction failed")
+      let result = try await ref.putDataAsync(data)
+      XCTAssertNotNil(result)
+    }
+
+    func testSimplePutFileNoMetadata() async throws {
+      let fileName = "hello&+@_ .txt"
+      let ref = storage.reference(withPath: "ios/public/" + fileName)
+      let data = try XCTUnwrap("Hello Swift World".data(using: .utf8), "Data construction failed")
+      let tmpDirURL = URL(fileURLWithPath: NSTemporaryDirectory())
+      let fileURL = tmpDirURL.appendingPathComponent("hello.txt")
+      try data.write(to: fileURL, options: .atomicWrite)
+      let result = try await ref.putFileAsync(from: fileURL)
+      XCTAssertNotNil(result)
+    }
+
+    func testSimpleGetData() async throws {
+      let ref = storage.reference(withPath: "ios/public/1mb2")
+      let result = try await ref.data(maxSize: 1024 * 1024)
+      XCTAssertNotNil(result)
+    }
+
+    func testSimpleGetDataWithTask() async throws {
+      let ref = storage.reference(withPath: "ios/public/1mb2")
+      let result = try await ref.data(maxSize: 1024 * 1024)
+      XCTAssertNotNil(result)
+    }
+
+    func testSimpleGetDataInBackgroundQueue() async throws {
+      actor MyBackground {
+        func doit(_ ref: StorageReference) async throws -> Data {
+          XCTAssertFalse(Thread.isMainThread)
+          return try await ref.data(maxSize: 1024 * 1024)
+        }
+      }
+      let ref = storage.reference(withPath: "ios/public/1mb2")
+      let result = try await MyBackground().doit(ref)
+      XCTAssertNotNil(result)
+    }
+
+    func testSimpleGetDataTooSmall() async {
+      let ref = storage.reference(withPath: "ios/public/1mb2")
+      do {
+        _ = try await ref.data(maxSize: 1024)
+        XCTFail("Unexpected success from getData too small")
+      } catch {
+        XCTAssertEqual((error as NSError).code, StorageErrorCode.downloadSizeExceeded.rawValue)
+      }
+    }
+
+    func testSimpleGetDownloadURL() async throws {
+      let ref = storage.reference(withPath: "ios/public/1mb2")
+
+      // Download URL format is
+      // "https://firebasestorage.googleapis.com:443/v0/b/{bucket}/o/{path}?alt=media&token={token}"
+      let downloadURLPattern =
+        "^https:\\/\\/firebasestorage.googleapis.com:443\\/v0\\/b\\/[^\\/]*\\/o\\/" +
+        "ios%2Fpublic%2F1mb2\\?alt=media&token=[a-z0-9-]*$"
+
+      let downloadURL = try await ref.downloadURL()
+      let testRegex = try NSRegularExpression(pattern: downloadURLPattern)
+      let urlString = downloadURL.absoluteString
+      XCTAssertEqual(testRegex.numberOfMatches(in: urlString,
+                                               range: NSRange(location: 0,
+                                                              length: urlString.count)), 1)
```

Comment:
This API might help make this a bit more concise (and help with formatting)
https://developer.apple.com/documentation/foundation/1408420-nsrangefromstring

Reply:
Hmm, I didn't find a way to use that one, but did switch to using `firstMatch` which seems cleaner.

Reply:
Hmm, I didn't find a way to use that one, but did switch to using firstMatch which seems cleaner.